internal/decision/engine.go

```diff
@@ -185,27 +191,36 @@ func newEngine(bs bstore.Blockstore, bstoreWorkerCount int, peerTagger PeerTagge
 	e := &Engine{
 		ledgerMap:                       make(map[peer.ID]*ledger),
 		scoreLedger:                     scoreLedger,
-		bsm:                             newBlockstoreManager(bs, bstoreWorkerCount),
+		bsm:                             newBlockstoreManager(ctx, bs, bstoreWorkerCount),
 		peerTagger:                      peerTagger,
 		outbox:                          make(chan (<-chan *Envelope), outboxChanBuffer),
 		workSignal:                      make(chan struct{}, 1),
 		ticker:                          time.NewTicker(time.Millisecond * 100),
 		maxBlockSizeReplaceHasWithBlock: maxReplaceSize,
-		taskWorkerCount:                 taskWorkerCount,
+		taskWorkerCount:                 engineTaskWorkerCount,
 		sendDontHaves:                   true,
 		self:                            self,
 		peerLedger:                      newPeerLedger(),
+		pendingGauge:                    metrics.NewCtx(ctx, "pending_tasks", "Total number of pending tasks").Gauge(),
+		activeGauge:                     metrics.NewCtx(ctx, "active_tasks", "Total number of active tasks").Gauge(),
 	}
 	e.tagQueued = fmt.Sprintf(tagFormat, "queued", uuid.New().String())
 	e.tagUseful = fmt.Sprintf(tagFormat, "useful", uuid.New().String())
 	e.peerRequestQueue = peertaskqueue.New(
 		peertaskqueue.OnPeerAddedHook(e.onPeerAdded),
 		peertaskqueue.OnPeerRemovedHook(e.onPeerRemoved),
 		peertaskqueue.TaskMerger(newTaskMerger()),
-		peertaskqueue.IgnoreFreezing(true))
+		peertaskqueue.IgnoreFreezing(true),
+		peertaskqueue.MaxOutstandingWorkPerPeer(maxOutstandingBytesPerPeer))
 	return e
 }
 
+func (e *Engine) updateMetrics() {
+	stats := e.peerRequestQueue.Stats()
```

Comment:
This function seems like it's getting called frequently and `peerRequestQueue.Stats()` takes out locks. We should probably either figure out a better way to do this (poll less frequently, or have a better locking strategy in the peerstaskqueue).

Reply:
Time-based polling is not good because it looses resolution during bursty times when you most need it. So instead I've modified metrics to be reported on each 100 operations.